pkgs/unified_analytics/lib/src/event.dart

```diff
@@ -716,11 +719,93 @@ final class Event {
       other is Event &&
       other.runtimeType == runtimeType &&
       other.eventName == eventName &&
-      compareEventData(other.eventData, eventData);
-
-  @override
-  String toString() => jsonEncode({
+      _compareEventData(other.eventData, eventData);
+
+  /// Converts an instance of [Event] to JSON.
+  ///
+  /// Example for [Event.timing] converted to JSON below.
+  /// ```json
+  /// {
+  ///   "eventName": "timing",
+  ///   "eventData": {
+  ///     "workflow": "my-work-flow",
+  ///     "variableName": "my-variable",
+  ///     "elapsedMilliseconds": 123,
+  ///     "label": "my-label"
+  ///   }
+  /// }
+  /// ```
+  String toJson() => jsonEncode({
         'eventName': eventName.label,
         'eventData': eventData,
       });
+
+  @override
+  String toString() => toJson();
+
+  /// Utility function to take in two maps [a] and [b] and compares them
+  /// to ensure that they have the same keys and values.
+  ///
+  /// Used for the equality operator.
+  bool _compareEventData(Map<String, Object?> a, Map<String, Object?> b) {
+    final keySetA = a.keys.toSet();
+    final keySetB = b.keys.toSet();
+    final intersection = keySetA.intersection(keySetB);
+
+    // Ensure that the keys are the same for each object
+    if (intersection.length != keySetA.length ||
+        intersection.length != keySetB.length) {
+      return false;
+    }
+
+    // Ensure that each of the key's values are the same
+    for (final key in a.keys) {
+      if (a[key] != b[key]) return false;
+    }
+
+    return true;
+  }
+
+  /// Returns a valid instance of [Event] if [json] follows the correct schema.
+  ///
+  /// Schema is below with the following notes:
+  /// - The `eventName` key must have a string value that exists in [DashEvent].
+  /// - The `eventData` key must have a nested object as its value.
+  /// ```json
+  /// {
+  ///   "eventName": "string-label",
+  ///   "eventData": {
+  ///     "myVar1": "value1",
+  ///     "myVar2": 123
+  ///   }
+  /// }
+  /// ```
+  static Event? fromJson(String json) {
+    try {
+      final jsonMap = jsonDecode(json) as Map<String, Object?>;
+
+      // Ensure the required keys are present
+      if (!jsonMap.containsKey('eventName') ||
+          !jsonMap.containsKey('eventData')) {
+        return null;
+      }
+
+      // Ensure the values for each key is the correct type
+      final eventName = jsonMap['eventName'];
+      final eventData = jsonMap['eventData'];
+      if (eventName is! String || eventData is! Map<String, Object?>) {
+        return null;
+      }
+
+      // Retrieve the correct DashEvent enum from the provided label
+      final dashEvent = DashEvent.getDashEventByLabel(eventName);
+      if (dashEvent == null) {
+        return null;
+      }
+
+      return Event._(eventName: dashEvent, eventData: eventData);
```

Comment:
I think you can simplify this using pattern matching:

```
if (jsonMap
    case  {
      'eventName': final String eventName, // you might be able to go further here to validate the DashEvent check in the pattern
      'eventData': final Map<String, Object?> eventData,
    }) {
  // Retrieve the correct DashEvent enum from the provided label
  final dashEvent = DashEvent.getDashEventByLabel(eventName);
  if (dashEvent == null) {
    return null;
  }
  return Event._(eventName: dashEvent, eventData: eventData);
}
```

Reply:
Oh nice, this is a clever, had to catch up on Craigs pattern matching video to understand the syntax but it looks like this use case is highlighted pretty much exactly!

https://www.youtube.com/watch?v=aLvlqD4QS7Y&ab_channel=Flutter

Thank you for the suggestion! :)

Reply:
(Still digging through if this is possible) but i was able to simplify by adding a `when` clause to the `case` as the below

```dart
  static Event? fromJson(String json) {
    try {
      final jsonMap = jsonDecode(json) as Map<String, Object?>;

      // Ensure that eventName is a string and a valid label and
      // eventData is a nested object
      if (jsonMap
          case {
            'eventName': final String eventName,
            'eventData': final Map<String, Object?> eventData,
          } when DashEvent.getDashEventByLabel(eventName) != null) {
        return Event._(
          eventName: DashEvent.getDashEventByLabel(eventName)!,
          eventData: eventData,
        );
      }

      return null;
    } on FormatException {
      return null;
    }
  }
```

But it seems like i would have to make a call to `DashEvent.getDashEventByLabel` twice, do you know if its possible to assign this variable AND have it get checked within the `when` clause?

If it's not possible, i'll probably remove the `when` clause and leave the check within the body of the `if`

Reply:
drive by comment--at least with this particular code, I think the solution is to pull `DashEvent.getDashEventByLabel(eventName)` into a local variable before the if statement.

Reply:
I wouldn't have access to that variable above the if statement though because it gets assigned in the `if-case` statement.. unless we don't check if it is a string first

Reply:
ahh yeah, I see what you mean. carry on :)

Reply:
Ah yeah I'm now sure if `DashEvent.getDashEventByLabel(eventName)` can be stored in a variable. Maybe @munificent knows?

Reply:
There isn't always a great way to interleave a series of pattern matches and guards so that you can incrementally bind stuff after a guard passes. In this case, you could do:

```dart
if (jsonMap
    case {
      'eventName': final String eventName,
      'eventData': final Map<String, Object?> eventData,
    }) {
  if (DashEvent.getDashEventByLabel(eventName) case var event?) {
    return Event._(eventName: event, eventData: eventData);
  }
}
```

Whether you like seeing nested if-case statements like this is a matter of taste. :)

If you wanted to really jam everything into a pattern, you *could* make `DashEvent.getDashEventByLabel` an extension on String:

```dart
extension DashEventExtensions on String {
  // TODO: Whatever the real return type should be.
  Object? get dashEventForLabel => ...
}
```

You can call extension getters in object patterns, so that would let you do:

```dart
if (jsonMap
    case {
      'eventName': String(dashEventForLabel: var event?),
      'eventData': final Map<String, Object?> eventData,
    }) {
  return Event._(
    eventName: event,
    eventData: eventData,
  );
}
```

Then the pattern will only match if the map's `'eventName'` key is mapped to a string where calling `dashEventForLabel` on it returns a non-null value. When that happens, the value is bound to `event`.

But this feels to me like twisting the language out of shape and I wouldn't do it (though it is kind of cool that you *can*).

Reply:
We've discussed making `case` an expression instead of directly built into the if-case statement. If we did that, then you could probably chain them using `&&` like:

```dart
if (jsonMap
    case {
      'eventName': final String eventName,
      'eventData': final Map<String, Object?> eventData,
    } &&
    DashEvent.getDashEventByLabel(eventName) case var event?) {
  return Event._(eventName: event, eventData: eventData);
}
```

But that's not in the language now and it's not clear if it's a good idea. It raises lots of tricky questions around the scope of the pattern variables.

Reply:
@munificent yeah looks like your example in the second comment would be ideal but i can see how it get tricky..  i went with just having a separate conditional within the body of the `if-case` so it's easy to understand, thank you for the thorough comment!